src/garage/torch/baselines/gaussian_mlp_baseline.py

```diff
@@ -0,0 +1,170 @@
+        optimizer (object): The optimizer of the algorithm. Should be the
+            optimizers in torch.optim.
+        optimizer_args (dict): Optimizer arguments.
+    """
+
+    def __init__(self,
+                 env_spec,
+                 hidden_sizes=(32, 32),
+                 hidden_nonlinearity=torch.tanh,
+                 hidden_w_init=nn.init.xavier_uniform_,
+                 hidden_b_init=nn.init.zeros_,
+                 output_nonlinearity=None,
+                 output_w_init=nn.init.xavier_uniform_,
+                 output_b_init=nn.init.zeros_,
+                 learn_std=True,
+                 init_std=1.0,
+                 min_std=1e-6,
+                 max_std=None,
+                 std_parameterization='exp',
+                 layer_normalization=False,
+                 normalize_observation=True,
+                 normalize_return=True,
+                 optimizer=torch.optim.Adam,
+                 optimizer_args=None):
+        super().__init__(env_spec)
+
+        self._module = GaussianMLPModule(
+            input_dim=env_spec.observation_space.flat_dim,
+            output_dim=1,
+            hidden_sizes=hidden_sizes,
+            hidden_nonlinearity=hidden_nonlinearity,
+            hidden_w_init=hidden_w_init,
+            hidden_b_init=hidden_b_init,
+            output_nonlinearity=output_nonlinearity,
+            output_w_init=output_w_init,
+            output_b_init=output_b_init,
+            learn_std=learn_std,
+            init_std=init_std,
+            min_std=min_std,
+            max_std=max_std,
+            std_parameterization=std_parameterization,
+            layer_normalization=layer_normalization)
+
+        self._normalize_observation = normalize_observation
+        self._normalize_return = normalize_return
+
+        if optimizer_args is None:
+            optimizer_args = dict()
+
+        self._optimizer = make_optimizer(optimizer, self._module,
+                                         **optimizer_args)
+
+    def fit(self, paths):
+        """Fit regressor based on paths.
+
+        Args:
+            paths (list[dict]): Sample paths.
+
+        """
+        # observations = xs, return = ys
+        observations = np.concatenate([p['observations'] for p in paths])
+        returns = np.concatenate([p['returns'] for p in paths])
+
+        if self._normalize_observation:
+            mean = np.mean(observations, axis=0, keepdims=True)
+            std = np.std(observations, axis=0, keepdims=True) + 1e-8
+            observations = (observations - mean) / std
+
+        if self._normalize_return:
+            mean = np.mean(returns, axis=0, keepdims=True)
+            std = np.std(returns, axis=0, keepdims=True) + 1e-8
+            returns = (returns - mean) / std
+
+        dist = self._module(torch.Tensor(observations))
+        ll = dist.log_prob(torch.Tensor(returns))
+        loss = -ll.mean()
+
+        self._optimizer.zero_grad()
+        loss.backward()
```

Comment:
in PPO, you need to run minibatches on this update, just as you run it on the policy.

likewise, in TRPO, you should use a ConjugateGradientOptimizer just as you use one for the policy.

one way to address this is to remove the fit/predict interface here and instead make the baseline just another module in your algorithm.

inside the algorithm, there are two optimization steps. first, you update the policy and keep the baseline fixed (torch.nograd or torch.detach). second, you update the baseline and keep the policy fixed.

Reply:
i suggest you get this PR merged and make the baseline a separate PR.